src/cryptonote_core/tx_pool.cpp

```diff
@@ -722,28 +730,45 @@ namespace cryptonote
   //TODO: investigate whether boolean return is appropriate
   bool tx_memory_pool::get_relayable_transactions(std::vector<std::tuple<crypto::hash, cryptonote::blobdata, relay_method>> &txs) const
   {
-    CRITICAL_REGION_LOCAL(m_transactions_lock);
-    CRITICAL_REGION_LOCAL1(m_blockchain);
+    std::vector<std::pair<crypto::hash, txpool_tx_meta_t>> change_timestamps;
     const uint64_t now = time(NULL);
     txs.reserve(m_blockchain.get_txpool_tx_count());
-    m_blockchain.for_all_txpool_txes([this, now, &txs](const crypto::hash &txid, const txpool_tx_meta_t &meta, const cryptonote::blobdata *){
+
+    CRITICAL_REGION_LOCAL(m_transactions_lock);
+    CRITICAL_REGION_LOCAL1(m_blockchain);
+    LockedTXN lock(m_blockchain.get_db());
+    m_blockchain.for_all_txpool_txes([this, now, &txs, &change_timestamps](const crypto::hash &txid, const txpool_tx_meta_t &meta, const cryptonote::blobdata *){
       // 0 fee transactions are never relayed
       if(!meta.pruned && meta.fee > 0 && !meta.do_not_relay)
       {
-        if (!meta.dandelionpp_stem && now - meta.last_relayed_time <= get_relay_delay(now, meta.receive_time))
-          return true;
-        if (meta.dandelionpp_stem && meta.last_relayed_time < now) // for dandelion++ stem, this value is the embargo timeout
-          return true;
+        const relay_method tx_relay = meta.get_relay_method();
+        switch (tx_relay)
+        {
+          case relay_method::stem:
+          case relay_method::forward:
+            if (meta.last_relayed_time > now)
+              return true; // continue to next tx
+            change_timestamps.emplace_back(txid, meta);
+            break;
+          default:
+          case relay_method::none:
```

Comment:
The `return true` after this chunk is within the callback given to `for_all_txpool_txes`, so it should continue to the next transaction. If you meant should the entire function unconditionally `return true`, there is a long standing `TODO` to investigate whether `true` should always be returned. I have yet to investigate this longstanding `TODO`.

Reply:
Sorry just got your comment, misread it previously. Yes, I will add an explicit `return` for the `none` label.

Reply:
Although to clarify - `get_relay_method()` returns `relay_method::none` when `meta.do_not_relay == true` which is already filtered above, but another explicit check cannot hurt.

Reply:
If it's a case that cannot happen, then it's fine. I meant just to avoid going into the subsequent age test. I did not realize this case was impossible in practice.

Reply:
I added a `return true` for both `default` and `relay_method::none` cases. It seemed worth adding for the former, so why not.